### ISSUE TITLE
Parent onRendered triggers before child's

### ISSUE THREAD

**Issue by ManuelDeLeon** [meteor/meteor#4410]
The short version is that Blaze.View.prototype.onViewReady uses Tracker.afterFlush to queue the rendered callbacks. This opens up the possibility of a parent's rendered callback to run before a child's one, even though the child already rendered.

This creates issues if you're processing elements on the onRendered functions. The typical example is "locate all unprocessed elements and do something with them, then flag them as processed". If onViewReady uses Tracker.afterFlush then a parent template can process an element before a child template can do it.

Here's a repro: https://github.com/ManuelDeLeon/blaze-rendering-issue

The hierarchy is:

```
body 
  -> editContact 
    -> card
    -> editContactForm
```

`editContact` only displays `card` if `editContactForm` has rendered.

When you run it the output is:

```
editContact.onCreated
editContactForm.onCreated
editContactForm.onRendered
card.onCreated
editContact.onRendered
============= editContact rendered with a card! =============
card.onRendered
```

In other words, `card` renders on the page but `card.onRendered` is called after `editContact.onRendered` (the parent).

The problem is that Tracker.afterFlush sets functions to run in the next cycle, but it allows other computations to be injected into the current queue while the queue is being processed .

The problem is fixed if you use a setTimeout instead of Tracker.afterFlush. The onRendered functions are still set to run on the next cycle but they don't allow other functions to be injected in the queue before they're processed.

So instead of:

```
Blaze.View.prototype.onViewReady = function (cb) {
  var self = this;
  var fire = function () {
    Tracker.afterFlush(function () {
      if (! self.isDestroyed) {
        Blaze._withCurrentView(self, function () {
          cb.call(self);
        });
      }
    });
  };
  self._onViewRendered(function onViewRendered() {
    if (self.isDestroyed)
      return;
    if (! self._domrange.attached)
      self._domrange.onAttached(fire);
    else
      fire();
  });
};
```

You would have:

```
Blaze.View.prototype.onViewReady = function (cb) {
  var self = this;
  var fire = function () {
    setTimeout(function () {
      if (! self.isDestroyed) {
        Blaze._withCurrentView(self, function () {
          cb.call(self);
        });
      }
    }, 0);
  };
  self._onViewRendered(function onViewRendered() {
    if (self.isDestroyed)
      return;
    if (! self._domrange.attached)
      self._domrange.onAttached(fire);
    else
      fire();
  });
};
```

The only change is that `Tracker.afterFlush` becomes `setTimeout`.

onCreated and onDestroyed could get the same treatment (haven't looked at how they're triggered).

Maybe there's another way of dealing with this issue.

**Comment by Slava** [comment]
Closing as a duplicate of https://github.com/meteor/meteor/issues/4166

**Comment by Slava** [comment]
Actually, it is related, but different. I don't have a good response on why this behavior happens.

**Comment by glasser** [comment]
Actually, I don't think this is a bug.  When the editContact template is rendered, card is not part of it, so it's valid to call the onRendered callback before card is ready.  card is added later without rerendering editContact.

**Comment by ManuelDeLeon** [comment]
@glasser That's not what's happening here. By the time editContact-onRendered is called, the card template is already rendered on the page. In the repro editContact-onRendered has an if to see if the card is already on the page, and it is (hence you see "==== editContact rendered with a card! ===").

**Comment by ManuelDeLeon** [comment]
@Slava @glasser  Please consider reopening this issue.

*"When the editContact template is rendered, card is not part of it"*
The problem is that between the time editContact is rendered (wihout card in it) and the editContact-onRendered function is executed, the card template is allowed to be inserted. So by the time editContact-onRendered is executed card is already part of it.

*"card is added later without rerendering editContact"*
card is added/rendered after editContact but before the editContact-onRendered executes.

If you run the repro you'll get the following order:
```
editContact -renders on the page-
card -renders on the page-
editContact-onRendered -executes-
card-onRendered -executes-
```

This creates problems is you're processing UI elements. Like the issue mentions, the typical example is "locate all unprocessed elements and do something with them, then flag them as processed". In this case card never gets a chance to process UI elements because it happens after editContact-onRendered executes with the (unprocessed) card elements in it.

Here are the two scenarios one would expect:

1) editContact-onRendered executes without a card in it.
```
editContact -renders on the page-
editContact-onRendered -executes-
card -renders on the page-
card-onRendered -executes-
```

2) editContact-onRendered executes with a card but card elements have already been processed by card-onRendered
```
editContact -renders on the page-
card -renders on the page-
card-onRendered -executes-
editContact-onRendered -executes-
```

**Comment by glasser** [comment]
I'm pretty sure I understood it correctly when I analyzed this a week ago.  Right now, it's hard for me to say because it's hard for me to fit this whole reproduction back into my head.

It was really hard for me to follow because you decided to use 7 main source files in 4 directories.  It was basically impossible to keep the whole reproduction in my head at once.  And the names used for the different pieces didn't really have anything to do with the issue.

If you rewrite the reproduction to have a single `html` file and a simple `template` file, and rename the templates to something like "outer", "inner", and "conditionallyIncluded", and update your last comment to use the new names, I might find the time to think through it again.  But right now it'll take me way too long to remember which piece of your reproduction is which.

I realize this wasn't obvious from our [bug report guidelines](https://github.com/meteor/meteor/wiki/Contributing-to-Meteor), so I've added something to hopefully make this clearer:

> Ideally, try to use as few source files as possible so that it's easy to see the whole reproduction on one screen, rather than making a large number of small files, even if that's not how you'd choose to structure an app

**Comment by mitar** [comment]
@glasser: I made a simple reproduction based on @ManuelDeLeon examples: https://github.com/mitar/meteor-issue4410

I can confirm that the problem exists: the child1's DOM exists before parent's `onRendered` is called, but child1's `onRendered` is called only afterwards:

```
child2 - onRendered
parent - onRendered
parent's onRendered called before child1's onRendered (even though child1 is already on the page)
child1 - onRendered
```

I agree this is something not expected and it should be fixed. So or DOM element should be inserted before parent's `onRendered` is called, or child1's `onRendered` should be called after DOM element is inserted and before parent's `onRendered` is called.

**Comment by glasser** [comment]
Thanks for the clean and easy to comprehend reproduction.

Still pretty sure this is working as intended: the parent fully renders without child1, and is never re-*rendered* (ie, fully recreated from scratch).  You'd see the same behavior if you added a giant timeout before the `Session.set`.

I do think that this makes the onRendered API not super useful, but it is working as intended.

I suspect that the idea of a component that can have its children mutated without being rerendered, while powerful, may not have been perfect.  I don't think that React (which Meteor supports) has this concern.  I'm not a frontend expert, though.

**Comment by mitar** [comment]
I think this is in fact related to this feature request #5354. So I think that the issue is that current Blaze code uses `afterFlush`. This means that it during an computation run computation itself gets invalidated, it reruns again immediately, and `afterFlush` gets to run just after that second computation. If instead `Tracker` would have an `afterRecompute` callback (proposed in #5354), then that would be a better place to put the `onRendered` callback and things would work correctly.

**Comment by glasser** [comment]
Hmm, you're right that something weird does happen due to the timing of the cycles.

But say we replaced the Session.set line with

```
    setTimeout(function () {
      console.log("setting var")
      Session.set("child2-onRendered-called", true)
    }, 500)
```

When I do that, I get 
```
child2 - onRendered
parent - onRendered
setting var
child1 - onRendered
```

ie, I don't get a second parent onRendered at all.  (Which I what I thought was happening before actually.)

So it's definitely kind of weird that the precise timing gives you a difference between one and two onRendered callback invocations.  But it's also definitely the case that Blaze isn't supposed to promise you a parent onRendered callback when its content is later mutated.

That said I'm mostly only active on this issue because I was trying to understand it better; I'm not enough of a Blaze expert to be part of a fix here.

**Comment by mitar** [comment]
> Still pretty sure this is working as intended: the parent fully renders without child1, and is never re-rendered (ie, fully recreated from scratch).

Hm, @glasser thank you for looking into it again, but I think we are not understanding each other. The issue is that there is some bad interaction between computation invalidation and callbacks. So it is good that parent is not re-rendered. The issue is that there is some DOM manipulation of the child1 happening at the seemingly strange time. Or:
* there should not be any child1 DOM present in the DOM tree when `onRendered` from parent is called (and just added afterwards)
* or, there should be child1 DOM present in the DOM tree, but then its `onRendered` should be called before parent's `onRendered`

Currently there is some DOM manipulation happening which one would not expect. The good pattern when doing Meteor Blaze programming is that you do not use outside code to modify your DOM. But you would never expect that Blaze itself would be modifying DOM tree so that it looks like it was modified from outside (if you are using `onRendered` to track what was done by Blaze).

**Comment by glasser** [comment]
Hi @mitar. I think this is beyond my ability to understand, but I'll reopen the issue.

**Comment by mitar** [comment]
> ie, I don't get a second parent onRendered at all. (Which I what I thought was happening before actually.)

No, that is not the problem. That is not what we are reporting in this issue and is good that it does not happen. With Blaze you get `onRendered` only the first time a template is rendered anyway. I think this is good.

>  But it's also definitely the case that Blaze isn't supposed to promise you a parent onRendered callback when its content is later mutated.

Yes, that is OK. We do not expect that `onRendered` is called twice. What we expect is that when parent `onRendered` is called, all children `onRendered` was called before and there is no extra dangling DOM there which for which was not `onRendered` called at some point.

I really think it is simply the issue of computation invalidating itself and thus triggering Blaze adding DOM, but callbacks are only called after the second rerun of the computation happened. Which is wrong.

> That said I'm mostly only active on this issue because I was trying to understand it better; I'm not enough of a Blaze expert to be part of a fix here.

Can you at least reopen it? :-) (edit: Thanks!)

**Comment by glasser** [comment]
Oh gosh, I did misread your very clear code when I suggested that parent onRendered was called twice.

**Comment by mitar** [comment]
(Just to give credit where credit is due, my "very clear code" is a direct copy-paste from @ManuelDeLeon's forum post here: https://forums.meteor.com/t/parent-templates-onrendered-called-before-childs/4373/7)

**Comment by mitar** [comment]
I made a fix for this in #5392.

**Comment by dgreensp** [comment]
I'm just getting up to speed on this issue.  The fact that, in general, child onRendered triggers before parent onRendered is undocumented but I guess happens in practice it happens reliably?  So then you have one child's onRendered trigger the rendering of another child... I'm not sure what I would expect to happen in this case.  It's kind of an oddball.

When does this come up?  It seems like an antipattern to thread view state through template rendering and out again.  In other words, based on this reactive data, we render these templates, and based on what Blaze ended up doing, we set this reactive data that causes this other rendering to happen...

**Comment by dgreensp** [comment]
Why does using `setTimeout` work?  Wouldn't `setTimeout` and `afterFlush` both execute `onRendered` in the same order that `setTimeout` or `afterFlush` was called?

**Comment by ManuelDeLeon** [comment]
@dgreensp 

`afterFlush` doesn't put it on the next js cycle, it appends the function to the current queue so it gets run on the same cycle.

`setTimeout` works because the new child is created on the next js cycle, after the parent is rendered (with its onRendered called).

**Comment by mitar** [comment]
I added some `console.log` outputs into `view.js` to show how things are happening. Numbers are computation IDs. With current core in Meteor:
```
doRender 1 body
end doRender 1 body
doRender 2 Template.parent
end doRender 2 Template.parent
dependency on session 3
doRender 4 if
end doRender 4 if
doRender 5 Template.child2
end doRender 5 Template.child2
onViewRendered Template.child2
onViewRendered Template.parent
onViewRendered body
adding afterFlush Template.child2
adding afterFlush Template.parent
adding afterFlush body
child2 - onRendered
invalidating session
dependency on session 3
doRender 4 if
dependency on session 6
doRender 7 with
end doRender 7 with
doRender 8 Template.child1
end doRender 8 Template.child1
onViewRendered Template.child1
adding afterFlush Template.child1
end doRender 4 if
parent - onRendered
parent's onRendered called before child1's onRendered (even though child1 is already on the page)
child1 - onRendered
```

So here we can see that because first `afterFlush` happens which invalidates the computation, because `Tracker` is written so that it processes one by one `afterFlush`, in the next `while` iteration it goes into running the computation again. Before parent `afterFlush` happens. So you have a situation where DOM is changed, but `child1` `afterFlush` callback is called much later.

With `setTimeout` approach:

```
doRender 1 body
end doRender 1 body
doRender 2 Template.parent
end doRender 2 Template.parent
dependency on session 3
doRender 4 if
end doRender 4 if
doRender 5 Template.child2
end doRender 5 Template.child2
onViewRendered Template.child2
onViewRendered Template.parent
onViewRendered body
adding setTimeout Template.child2
adding setTimeout Template.parent
adding setTimeout body
child2 - onRendered
invalidating session
parent - onRendered
dependency on session 3
doRender 4 if
dependency on session 6
doRender 7 with
end doRender 7 with
doRender 8 Template.child1
end doRender 8 Template.child1
onViewRendered Template.child1
adding setTimeout Template.child1
end doRender 4 if
child1 - onRendered
```

This one works because it breaks the fact that Tracker processes only one `afterFlush` at a time. It effectively pushes queue onto the runtime events queue, allowing both of callbacks to finish before the next flush cycle happens.

With my `afterRun` approach:

```
doRender 1 body
end doRender 1 body
doRender 2 Template.parent
end doRender 2 Template.parent
dependency on session 3
doRender 4 if
end doRender 4 if
doRender 5 Template.child2
end doRender 5 Template.child2
onViewRendered Template.child2
onViewRendered Template.parent
onViewRendered body
adding afterFlush Template.child2 null
adding afterFlush Template.parent null
adding afterFlush body null
child2 - onRendered
invalidating session
dependency on session 3
doRender 4 if
dependency on session 6
doRender 7 with
end doRender 7 with
doRender 8 Template.child1
end doRender 8 Template.child1
onViewRendered Template.child1
adding afterRun Template.child1 4
end doRender 4 if
child1 - onRendered
parent - onRendered
```

Here initially we use `afterFlush` for `child2` and `parent` because we are not inside a computation, but for `child1` we are, so we make sure we call a callback using `afterRun`. This makes it sure that it is called before `afterFlush` from the parent. In some it synces up DOM modification with the callback, assuring that when DOM changes callback is called immediately.

In fact the main issue here is that when using `afterFlush` a starvation is happening. In this case starvation of `onRendered` callbacks, in reactive-publish case starvation of publish updates to the client.

I think the same issue was observed in the past with Tracker and this is why breaking a 1000 loop was introduced. But what I am saying that that arbitrary breaking is not enough. We should also have an API to say that something should run after every computation (so that it does not starve) or that something is less critical and can run in free time after all reruns.

**Comment by dgreensp** [comment]
I'm remembering the mechanics of the situation now.  (I know you guys worked this out, just recalling it myself and writing it down more explicitly.)  Tracker processes two types of items at flush time, each with its own queue: Computation re-runs and afterFlushes.  It does this by repeatedly: drawing an item from the Computation queue if possible, otherwise drawing an item from the afterFlush queue.  Therefore, if an afterFlush invalidates a Computation, the Computation runs before any and all queued afterFlushes.  However, if that Computation adds an afterFlush, the afterFlush goes to the end of the afterFlush queue and executes last.  By the way, I don't remember any particular motivation for programming Tracker that way; it was just an arbitrary choice at the time.

That background is mostly needed to explain how the "fixes" work rather than how Blaze currently works.  To understand how the current behavior works, you just need to know that onRendered callbacks go into a queue, and if an onRendered callback indirectly leads to another onRendered callback, it gets added to the end of the queue of onRendered callbacks to call.

If I were to classify this situation as part of a larger pattern, it would be the fact that in Blaze and Tracker, there isn't any ordering or sorting of Computations or afterFlushes based on the parent-child view relationship.  This comes up, for example, when a parent view and a child view are independently invalidated.

Even in the particular case of onRendered callbacks, however, I don't think hacking the way they are queued really achieves the goal of making sure callbacks on children fire before callbacks on parents.  The assumption seems to be that newly-fired onRendered callbacks and old, queued onRendered callbacks have a child-parent relationship to each other -- but if you're changing reactive state in T.onRendered (that affects views outside of T), who's to say what parts of the view tree that affects?

I'd view this as a new feature, first of all, and talk about why it's important.  It seems like there are two things going on in the original use case -- using onRendered to process elements, and onRendered causing a state change that triggers more rendering.  I think the second half of that is just something that happens in UIs sometimes, and the real feature request is a way to ensure that onRendered callbacks are correctly ordered, regardless of how they occur.  So I would focus on implementing that feature in a way where it's clear that the desired property is achieved.

**Comment by brucejo75** [comment]
Any progress on this issue?

Luckily, I ran onto this issue while investigating why my onRendered call backs were being called before my template had rendered.

I am using a jquery validate package, groveling through the code I found that the dom was not accurate when the validation was called from within onRendered, so selector lookup was fubar.  I added a setTimeout and the dom is now accurate.  I also noticed that previously rendered template onRendered callbacks were being called, which was pretty confusing.

I do not know if the suggested solution is the right one, but I do believe you have a real bug on your hands.  And if Mitar is right that you are starving the onRendered callbacks that seems to be a problem, no?
